### ISSUE TITLE
Include Rangy?

### ISSUE THREAD

**Issue by Gankra** [OscarGodson/EpicEditor#239]
The two biggest issues, as far as I'm concerned, are the selection api and #100. Whenever I look up how to handle all the nasty corner-cases involved in manipulating contenteditables, I invariably find a post by the creator of Rangy, which is a library intended to smooth over the bumps in the various implementations of the Selection API. This guy clearly understands how this stuff is supposed to work.

In particular Rangy 1.3 should introduce this TextRange module: https://code.google.com/p/rangy/wiki/TextRangeModule

Which lets you abstract away all the nasty DOM stuff and do things like "extend the current selection by two visible characters" (as opposed to offsets in nodes). It also has a module that handles saving selections which could help with the html-pasting hack.

Basically, I'm just curious if you would even be open to adding Rangy as a dependency, before I waste a few days digging into proof-of-concept integration of it.

Misc Details:
- MIT license
- 1.3 is only in alpha (!) 

Size:
- Uncompressed core: 144kb
- Uncompressed core+modules: ~300kb
- Minified core: 46kb
- Minified core+modules: ~90kb

**Comment by OscarGodson** [comment]
This is tough. I wouldn't want to put this in EpicEditor core because of it's size. _Just_ rangy minified is more than EpicEditor + Marked + Editor Theme (unminified) + Base theme (unminified) + Preview theme (unminified). That's kinda crazy. The point of EpicEditor has always been to have a fully featured, easily scriptable, editor that isn't massive in size. This will double the minified source unfortunately.

This could very well be a good first extension/module for EpicEditor tho. I haven't spent much time on how to handle modules, but considering its size and the low usage it'll have in core itself for private use this sounds like a perfect example. A Rangy EpicEditor module which would give you that spec in that gist to make it super easy even for designers to add a button bar or shortcuts to trigger certain actions on selected text. 

What do you think about handling #33 + integrating this and we could setup a module's section on the site or something?

**Comment by Gankra** [comment]
Totally agree, the size is a big issue. Part of that is that they support far more than us. Including it as a module of EE would definitely make sense, except then we'd still have to resolve #100 some other way, since that shouldn't be a problem that exists in core.

**Comment by OscarGodson** [comment]
@Gankro yeah, there's work arounds to that tho that wont rely on selection stuff. I think TinyMCE does some super hack thing where, on paste, it creates a textarea, pastes in there, get's textarea value, and then appends that text to the contenteditable. I just haven't had time to try out that method.

**Comment by Gankra** [comment]
That method only works if you paste by keyboard, as far as I know

**Comment by OscarGodson** [comment]
@Gankro works for me on Chrome and Firefox with keyboard or right click -> paste: http://jsbin.com/egalug/2/edit

**Comment by Gankra** [comment]
Yeah it's the redirect text area thing that doesn't work with the other menu. It involves moving the cursor on keyed ctrl+v to a hidden text area.

Anyway you should check out the patch I mentionned in #100, it works with any paste method ad far as I can tell (but will only work for firefox when they make their next release-- hence why it wasn't in use).
